refactor(ast): box `TSThisParameter`

overlookmotel opened the pull request:
Box `TSThisParameter` in `Function`, `TSMethodSignature` and `TSFunctionType`. I assume `function(this: Whatever) {}` is a fairly rare syntax in TS code, and obviously never occurs in JS code, so it takes up a lot of space in `Function` for this uncommon case.

This change reduces `Function` from 136 bytes to 104.

overlookmotel commented:
* **#5325** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5325?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

DonIsaac commented:
I'm not seeing much of a change on codspeed. Is this due to the change itself or a deficiency in our benchmarks?

overlookmotel commented:
@DonIsaac You willing to approve this, and merge it?

DonIsaac commented:
> Probably function nodes are fairly rare compared to other nodes, so it will perform better for `Function`s, but there aren't enough of them for it to be noticeable on benchmarks.

I think checker.ts has a ton of functions, no?

overlookmotel committed:
refactor(ast): box `TSThisParameter` (#5325)

Box `TSThisParameter` in `Function`, `TSMethodSignature` and `TSFunctionType`. I assume `function(this: Whatever) {}` is a fairly rare syntax in TS code, and obviously never occurs in JS code, so it takes up a lot of space in `Function` for this uncommon case.

This change reduces `Function` from 136 bytes to 104.
